src/rbac/role_manager.rs

```diff
@@ -0,0 +1,16 @@
+pub trait RoleManager {
+    fn clone_box(&self) -> Box<dyn RoleManager>;
```

Comment:
Rather than require this method, you can add a constraint on the self type of this trait to require implementors are `Clone`.

Reply:
I learned this technical from https://users.rust-lang.org/t/how-to-clone-a-boxed-closure/31035  I don't know how to add clone constraint to self, can you be more specific?

Reply:
If you use `dyn Rolemanager` any where else you can not change this because your trait becomes object *unsafe* and you then can *not* refer to it as `dyn Rolemanager`. 
If you don't need that ability it would look like this
```rust 
// this means that every type that impl's RoleManager must be Clone
pub trait RoleManager: Clone {
//  No more clone_box
    fn clear(&mut self);
    fn print_roles(&self);
}

#[derive(Clone)]
struct Try(Option<fn()>);
impl RoleManager for Try {
    fn clear(&mut self) {}
    fn print_roles(&self) {}
}
```
[try it here](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=364e8ac75cbb0fc200f4c5f4b74cc47b)

Reply:
I tried to use `RoleManager: Clone` to replace boxed trait with clone, but gave up in the end, because this approach introduces so many trait constraints. Can you help make a pull request on this?

Reply:
You do use RoleManager as a trait object constraints that require knowing the Size of the trait object is what makes it unsafe. I should have looked around a bit more 
```rust
pub fn generate_g_function(rm: Box<dyn RoleManager>) -> Box<dyn MatchFnClone> 
```
So it is probably not worth changing just to remove one method that is perfectly fine. Although if you wanted to [this](https://users.rust-lang.org/t/object-safety-and-cloning-traits/22620/5?u=devinr528) an Rc may work I can try if you want or using dtonlay's crate or even easier [objekt](https://github.com/kardeiz/objekt-clonable) but it probably isn't worth it I don't think clone_box is all that much more expensive.